src/collapse/dtrace.rs

```diff
@@ -221,83 +231,108 @@ impl Folder {
             let (mut index, mut nstacks) = (0, 0);
 
-            // The loop...
+            // Loop through the input data in order to chunk it up and send it off to the worker threads...
             loop {
-                // First, read some data into the `utils::Buffer`.
+                // First, read some data into the buffer.
                 let n = reader.read_until(b'\n', &mut buf)?;
 
                 // If we're at the end of the data...
                 if n == 0 {
                     // Send the last slice and exit the loop.
-                    tx_input.send(Some(buf)).unwrap();
+                    let _ = tx_input.send(Some(buf));
                     break;
                 }
 
                 let line = &buf[index..index + n];
                 index += n;
 
+                // If we're at the end of a stack...
                 if is_end_of_stack(line) {
-                    // It's the end of a stack; count it
+                    // Count it.
                     nstacks += 1;
                     // If we've seen enough stacks to make up a slice...
                     if nstacks == self.nstacks_per_job {
                         // Send it.
                         let buf_capacity = buf.capacity();
                         let chunk = mem::replace(&mut buf, Vec::with_capacity(buf_capacity));
-                        tx_input.send(Some(chunk)).unwrap();
+                        if let Err(_) = tx_input.send(Some(chunk)) {
+                            // If this send returns an error, it means the children have shutdown
+                            // because one of them produced an error and has already sent
+                            // it to the main thread; so we should stop doing unnecessary work,
+                            // i.e. we should break.
+                            break;
+                        }
                         // Reset the state; mark the beginning of the next slice.
                         index = 0;
                         nstacks = 0;
                     }
                 }
             }
 
-            // Send shutdown signal.
+            // We've finished sending the input; so send the "no more input" signal,
+            // i.e. `None`, to the children.
             for _ in &handles {
-                tx_input.send(None).unwrap();
+                if let Err(_) = tx_input.send(None) {
+                    // If this send returns an error, it means the children have shutdown
+                    // because one of them has already produced an error and sent it to the
+                    // main thread; so we should break here.
+                    break;
+                }
             }
 
-            // Retrieve error, if any.
             drop(tx_output);
-            if let Ok(result) = rx_output.recv() {
-                result?;
-            }
+            let maybe_error = rx_output.iter().find(|result| result.is_err());
```

Comment:
Isn't all that is ever sent on `rx_output` an error? Is the `find` really necessary? In fact, why are we sending `Result<>` there at all, as opposed to just sending the errors without wrapping?